db/range_del_aggregator.cc

```diff
@@ -161,25 +163,25 @@ class CollapsedRangeDelMap : public RangeDelMap {
       assert(std::next(iter_) != rep_.end());
       assert(iter_->second != 0);
       RangeTombstone tombstone;
-      tombstone.start_key_ = iter_->first;
-      tombstone.end_key_ = std::next(iter_)->first;
+      tombstone.start_key_ = ExtractUserKey(iter_->first);
+      tombstone.end_key_ = ExtractUserKey(std::next(iter_)->first);
```

Comment:
This seems suspicious to me. Internally, the `CollapsedRangeDelMap` represents tombstones with internal-keys, but here you're truncating the tombstone again to be represented with user-keys. That will cause the start-key to cover more of the key space and end-key to cover less of it. Is this a problem? I'm not sure. I think this interface is only used when building an sstable and during compactions. During compactions, a truncated tombstone will only occur when the tombstone straddles two sstables. If the two sstables are an atomic unit (straddled by a user key) then the tombstone from the second sstable should blend with this one. If it is only the tombstone which straddles the sstable, then the truncation will have occurred on the tombstone sentinel key and everything should be ok as well. I think. This deserves additional thinking, testing, and commentary.

Reply:
Thanks for making us think about this more. We think it'd be a problem cases like this:

- SST1 at L1 contains `Put`s at a@50, b@40, and b@30
- SST2 at L1 contains `Put`s at b@20 and c@10
- A `DeleteRange` from a to c at seqnum 45 is stored (untruncated) in SST1 and SST2
- SST1 and SST2 are picked for L1->L2 compaction, along with some L2 files.
- The first output is SST3, which ends at b@40. At this point the `CompactionIterator`'s L1 `LevelIterator` is positioned at b@30, which is still in SST1. That means SST2 hasn't been opened so its range tombstone haven't been added to the `RangeDelAggregator`. Then we would simply be writing out a range tombstone with end_key b to SST3, which would not cover b@40, causing that key to reappear.

Reply:
Now we are learning the appeal of passing the atomic compaction unit boundaries as smallest and largest keys for `AddTombstones` :(.

Reply:
I suppose we can combine the two approaches. Use internal-keys for the endpoints of range tombstones in `RangeDelAggregator` and pass the compaction boundaries as the largest and smallest keys to `AddTombstones` when creating the `RangeDelAggregator` used for compaction. Looking at the code, it appears to be a bit irritating to plumb the compaction boundaries down to the `AddTombstones` call, but `CompactionJob` creates the `RangeDelAggregator` and it could call a constructor or some other method that says "uses these boundaries for truncation and ignore the boundaries passed to `AddTombstones`".

Reply:
@petermattis - yup, we were planning a hybrid approach, too. Should've written it here, sorry.

I also thought a bit about having overrides in `RangeDelAggregator`. My feeling is the boundaries we pass to `AddTombstones` should be specific to the file's level, though. The high-level concern is when the compaction's output level extends the compaction's range far to the right, it could cause range tombstones on the input level to extend arbitrary rightwards after compaction happens. I should come up with a specific example though.

Reply:
Actually, now I am seeing problems even with using the level's compaction boundaries as smallest/largest. Will try to make this more concrete after dinner :p, but the high-level problem is when a level's files are not all overlapping, then clamping based on atomic unit boundaries would give tighter tombstones than clamping based on compaction level boundaries.

Reply:
Example as promised:

- LSM has 7 levels, so L6 is bottommost
- L5 has two SSTs: SST1 has `Put`s from a to c. SST2 has `Put`s from d to f. A range tombstone from b to e is stored in both SSTs. Note SST1's `largest_key` is extended to d@kMaxSeqnum due to this range tombstone.
- L6 has two SSTs: SST3 has `Put`s from b to d. SST4 has `Put`s from d to f.
- L5 needs compaction. An L5->L6 compaction is chosen. Let's say SST2 is chosen from the input level, so the overlapping files chosen from the output level are SST3 and SST4. Since this is a compaction to the bottommost level, the keys from SST2 in the range [d, e) that are newer than the range tombstone will be written out with seqnum zero.
- The L5->L6 compaction finishes, creating SST5 spanning b to f.
- New data comes in the range [d, f], making its way down through memtable, L0, ..., L5. Now we have SST6 in L5.
- L5 needs compaction again. An L5->L6 compaction is chosen. This time, let's say SST1 was chosen. It overlaps with SST5 in L6. Now we have an optimization that goes back to the input level and expands the compaction to include any other files that are fully contained in the compaction's key-range. This pulls in SST6.
- The compaction runs, the keys in SST5 in the range [d, e) that were originally newer than the range tombstone get nuked, and now we have data loss.

Sorry, @abhimadan, for advocating solving this problem in `RangeDelAggregator` -- I wasn't able to foresee (and am still quite surprised by) all these intricate edge cases. My feeling now is our best hope is Peter's original proposal of precomputing atomic compaction unit boundaries, and using those for `AddTombstones`'s smallest/largest. What do you think?

Reply:
Wow, that's a complex edge case. Thanks for catching that!

Actually, I'm going to propose another hybrid approach here: rather than precomputing atomic compaction boundaries, we compute the atomic compaction boundaries at compaction time. Then we won't have to worry about extra bookkeeping in `FileMetaData`. After these boundaries are computed, plumbing them to `AddTombstones` is quite similar to the work I've already done on using "full" compaction boundaries. While we could potentially cache them in `FileMetaData` as suggested, I'm going to argue that it's not necessary, because the files in the input level will no longer exist after the compaction, and the files in the output level will have changed after the compaction, and so their `FileMetaData` will need to be updated accordingly.

I like this idea since it's simple and (hopefully) cheap to compute on every compaction, since we can just do this work while we iterate through the `FileMetaData` in a level. (On the other hand it's quite late at the time of writing and I may have missed something important.) Thoughts?